File: android/feature/auth/src/main/kotlin/com/dualmusic/feature/auth/AuthViewModel.kt
package com.dualmusic.feature.auth

import com.dualmusic.domain.geo.Country
import com.dualmusic.domain.geo.Countries

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dualmusic.domain.api.DomainError
import com.dualmusic.domain.auth.AuthUser
import com.dualmusic.domain.auth.RegisterRequest
import com.dualmusic.domain.user.UpdateProfileRequest
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** État global de session. */
sealed interface AuthState {
    data object Loading : AuthState          // réhydratation au démarrage
    data object SignedOut : AuthState
    /** Inscrit mais email non vérifié : on propose l'écran de saisie du code. */
    data class PendingEmailVerification(val user: AuthUser, val email: String) : AuthState
    /** Email vérifié : on demande les informations de profil (nom, pays, numéro). */
    data class PendingProfileCompletion(val user: AuthUser) : AuthState
    data class SignedIn(val user: AuthUser) : AuthState
}

/** Écran actif de la zone d'authentification. */
enum class AuthMode { LOGIN, REGISTER, FORGOT, RESET }

/** État de l'écran auth (connexion / inscription / mot de passe oublié / reset). */
data class SignInUiState(
    val mode: AuthMode = AuthMode.LOGIN,
    val email: String = "",
    val password: String = "",
    val confirmPassword: String = "",
    val fullName: String = "",
    val phone: String = "",
    val birthDate: String = "",
    val gender: String = "",
    val country: Country = Countries.DEFAULT,
    val referralCode: String = "",
    val acceptTerms: Boolean = false,
    // Reset de mot de passe.
    val resetCode: String = "",
    val newPassword: String = "",
    // Vérification email (écran dédié).
    val verifyCode: String = "",
    val isSubmitting: Boolean = false,
    val error: String? = null,
    val info: String? = null,
) {
    /** Le formulaire courant est-il soumettable ? (règles = celles du backend). */
    val canSubmit: Boolean
        get() = !isSubmitting && when (mode) {
            AuthMode.LOGIN -> email.contains("@") && password.length >= 8
            // Étape 1 : email + mot de passe + confirmation + acceptation obligatoire des
            // documents légaux. Les autres infos (nom, pays, numéro) sont demandées APRÈS
            // validation du code email.
            AuthMode.REGISTER ->
                email.contains("@") && password.length >= 8 && password == confirmPassword && acceptTerms
            AuthMode.FORGOT -> email.contains("@")
            AuthMode.RESET -> email.contains("@") && resetCode.length in 4..8 && newPassword.length >= 8
        }
}

/**
 * ViewModel de l'authentification (MVI léger).
 *
 * Gère la connexion, l'inscription (mêmes champs que le web), le mot de passe oublié
 * (demande + reset par code email) et la vérification de l'email par code.
 */
class AuthViewModel(private val repository: AuthRepository) : ViewModel() {

    private val _authState = MutableStateFlow<AuthState>(AuthState.Loading)
    val authState: StateFlow<AuthState> = _authState.asStateFlow()

    private val _uiState = MutableStateFlow(SignInUiState())
    val uiState: StateFlow<SignInUiState> = _uiState.asStateFlow()

    fun onEmailChange(v: String) = _uiState.update { it.copy(email = v, error = null) }
    fun onPasswordChange(v: String) = _uiState.update { it.copy(password = v, error = null) }
    fun onConfirmPasswordChange(v: String) = _uiState.update { it.copy(confirmPassword = v, error = null) }
    fun onFullNameChange(v: String) = _uiState.update { it.copy(fullName = v, error = null) }
    fun onPhoneChange(v: String) = _uiState.update { it.copy(phone = v, error = null) }
    fun onBirthDateChange(v: String) = _uiState.update { it.copy(birthDate = v, error = null) }
    fun onGenderChange(v: String) = _uiState.update { it.copy(gender = v, error = null) }
    fun onReferralChange(v: String) = _uiState.update { it.copy(referralCode = v.uppercase(), error = null) }
    fun onCountrySelected(country: Country) = _uiState.update { it.copy(country = country, error = null) }
    fun onAcceptTermsChange(v: Boolean) = _uiState.update { it.copy(acceptTerms = v, error = null) }
    fun onResetCodeChange(v: String) = _uiState.update { it.copy(resetCode = v.filter { c -> c.isDigit() }, error = null) }
    fun onNewPasswordChange(v: String) = _uiState.update { it.copy(newPassword = v, error = null) }
    fun onVerifyCodeChange(v: String) = _uiState.update { it.copy(verifyCode = v.filter { c -> c.isDigit() }, error = null) }

    /** Change de mode (connexion/inscription/oublié/reset), en nettoyant messages. */
    fun setMode(mode: AuthMode) = _uiState.update { it.copy(mode = mode, error = null, info = null) }

    /** Connexion via ID token Google (obtenu par Credential Manager côté écran). */
    fun signInWithGoogle(idToken: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSubmitting = true, error = null) }
            runCatching { repository.loginWithGoogle(idToken) }
                .onSuccess { _authState.value = AuthState.SignedIn(it.user) }
                .onFailure { e -> _uiState.update { it.copy(error = e.message ?: com.dualmusic.core.ui.i18n.appStrings.errGoogleSignInFailed) } }
            _uiState.update { it.copy(isSubmitting = false) }
        }
    }

    /** Remonte une erreur du sélecteur Google (annulation, pas de compte…). */
    fun onGoogleError(message: String?) = _uiState.update { it.copy(error = message) }

    /** Réhydrate la session au lancement (via le refresh token persisté). */
    fun bootstrap() {
        viewModelScope.launch {
            _authState.value = runCatching { AuthState.SignedIn(repository.me().user) }
                .getOrDefault(AuthState.SignedOut)
        }
    }

    /** Soumet le formulaire courant selon le [SignInUiState.mode]. */
    fun submit() {
        val s = _uiState.value
        if (!s.canSubmit) return
        viewModelScope.launch {
            _uiState.update { it.copy(isSubmitting = true, error = null, info = null) }
            when (s.mode) {
                AuthMode.LOGIN -> runCatching { repository.login(s.email.trim(), s.password) }
                    .onSuccess { _authState.value = AuthState.SignedIn(it.user) }
                    .onFailure { setError(it) }

                // Étape 1 : on ne crée le compte qu'avec email + mot de passe. Le backend
                // envoie automatiquement le code de vérification par email.
                AuthMode.REGISTER -> runCatching {
                    repository.register(RegisterRequest(email = s.email.trim(), password = s.password))
                }.onSuccess {
                    _authState.value = AuthState.PendingEmailVerification(it.user, s.email.trim())
                }.onFailure { setError(it) }

                AuthMode.FORGOT -> runCatching { repository.forgotPassword(s.email.trim()) }
                    .onSuccess {
                        _uiState.update {
                            it.copy(mode = AuthMode.RESET, info = "Si un compte existe, un code a été envoyé par email.")
                        }
                    }
                    .onFailure { setError(it) }

                AuthMode.RESET -> runCatching {
                    repository.resetPassword(s.email.trim(), s.resetCode, s.newPassword)
                }.onSuccess {
                    _uiState.update {
                        it.copy(mode = AuthMode.LOGIN, password = "", info = "Mot de passe réinitialisé. Connecte-toi.")
                    }
                }.onFailure { setError(it) }
            }
            _uiState.update { it.copy(isSubmitting = false) }
        }
    }

    /** Vérifie le code email ; en cas de succès, passe à la saisie des infos de profil. */
    fun verifyEmail() {
        val pending = _authState.value as? AuthState.PendingEmailVerification ?: return
        val code = _uiState.value.verifyCode
        if (code.length !in 4..8) return
        viewModelScope.launch {
            _uiState.update { it.copy(isSubmitting = true, error = null) }
            runCatching { repository.verifyEmailOtp(code) }
                .onSuccess { _authState.value = AuthState.PendingProfileCompletion(pending.user) }
                .onFailure { setError(it) }
            _uiState.update { it.copy(isSubmitting = false) }
        }
    }

    /**
     * Étape 3 : enregistre les informations de profil (nom, pays, numéro) puis entre dans
     * l'app. Le numéro est optionnel ; le nom est requis.
     */
    fun completeProfile() {
        val pending = _authState.value as? AuthState.PendingProfileCompletion ?: return
        val s = _uiState.value
        if (s.fullName.isBlank()) {
            _uiState.update { it.copy(error = com.dualmusic.core.ui.i18n.appStrings.errNameRequired) }
            return
        }
        viewModelScope.launch {
            _uiState.update { it.copy(isSubmitting = true, error = null) }
            runCatching {
                repository.updateProfile(
                    UpdateProfileRequest(
                        fullName = s.fullName.trim(),
                        countryCode = s.country.code,
                        phone = s.phone.trim().ifBlank { null },
                        phoneCountryCode = s.country.dial,
                        birthDate = s.birthDate.trim().ifBlank { null },
                        gender = s.gender.ifBlank { null },
                    ),
                )
            }.onSuccess { _authState.value = AuthState.SignedIn(pending.user) }
                .onFailure { setError(it) }
            _uiState.update { it.copy(isSubmitting = false) }
        }
    }

    /** Passe la saisie des infos de profil (pourra les compléter plus tard). */
    fun skipProfile() {
        val pending = _authState.value as? AuthState.PendingProfileCompletion ?: return
        _authState.value = AuthState.SignedIn(pending.user)
    }

    /** (Ré)envoie le code de vérification email. */
    fun resendEmailCode() {
        viewModelScope.launch {
            runCatching { repository.sendEmailOtp() }
                .onSuccess { _uiState.update { it.copy(info = "Nouveau code envoyé.") } }
                .onFailure { setError(it) }
        }
    }

    /** Passe la vérification (non bloquante) → saisie des infos de profil. */
    fun skipVerification() {
        val pending = _authState.value as? AuthState.PendingEmailVerification ?: return
        _authState.value = AuthState.PendingProfileCompletion(pending.user)
    }

    /** Déconnexion. */
    fun signOut() {
        viewModelScope.launch {
            repository.logout()
            _authState.value = AuthState.SignedOut
            _uiState.value = SignInUiState()
        }
    }

    private fun setError(t: Throwable) = _uiState.update { it.copy(error = friendlyMessage(t)) }

    /** Traduit une erreur backend en message utilisateur. */
    private fun friendlyMessage(t: Throwable): String = when {
        t is DomainError && t.httpStatus == 401 -> "Email ou mot de passe incorrect."
        t is DomainError && t.httpStatus == 409 -> "Cet email est déjà utilisé."
        t is DomainError && (t.code == "OTP_INVALID" || t.code == "OTP_EXPIRED") -> "Code invalide ou expiré."
        t is DomainError && t.code == "VALIDATION_ERROR" -> "Champs invalides (email valide + mot de passe ≥ 8 caractères)."
        t is DomainError && t.isRetriable -> "Connexion instable. Réessaie."
        t is DomainError -> t.message
        else -> "Une erreur est survenue. Réessaie."
    }
}

File: ios/Packages/DualMusicKit/Sources/FeatureAuth/AuthViewModel.swift
import Foundation
import Observation
import CoreNetwork
import CoreUI
import DomainModels

/// État global de session — pilote la racine de l'app.
public enum AuthState: Equatable {
    /// Réhydratation au démarrage.
    case loading
    case signedOut
    /// Inscrit mais email non vérifié : on propose l'écran de saisie du code.
    case pendingEmailVerification(user: AuthUser, email: String)
    /// Email vérifié : on demande les informations de profil (nom, pays, numéro).
    case pendingProfileCompletion(user: AuthUser)
    case signedIn(user: AuthUser)

    public static func == (lhs: AuthState, rhs: AuthState) -> Bool {
        switch (lhs, rhs) {
        case (.loading, .loading), (.signedOut, .signedOut): return true
        case let (.pendingEmailVerification(a, ae), .pendingEmailVerification(b, be)): return a.id == b.id && ae == be
        case let (.pendingProfileCompletion(a), .pendingProfileCompletion(b)): return a.id == b.id
        case let (.signedIn(a), .signedIn(b)): return a.id == b.id
        default: return false
        }
    }
}

/// Écran actif de la zone d'authentification.
public enum AuthMode: Equatable {
    case login
    case register
    case forgot
    case reset
}

/// ViewModel de l'authentification (MVI léger) — miroir de `AuthViewModel` Android.
///
/// Gère la connexion, l'inscription en 3 étapes (compte → code email → profil), le mot de
/// passe oublié (demande + reset par code) et la connexion Google native.
///
/// Le fournisseur d'ID token Google est **injecté** (``googleIdTokenProvider``) : le SDK
/// Google reste dans la cible applicative, ce module ne dépend d'aucun tiers.
@Observable
@MainActor
public final class AuthViewModel {

    // MARK: État de session

    /// État global (piloté par la racine de l'app).
    public private(set) var authState: AuthState = .loading

    // MARK: État du formulaire

    public var mode: AuthMode = .login
    public var email: String = ""
    public var password: String = ""
    public var confirmPassword: String = ""
    public var fullName: String = ""
    public var phone: String = ""
    public var country: Country = Countries.default
    public var referralCode: String = ""
    /// Code de réinitialisation reçu par email (normalisé en chiffres par la vue).
    public var resetCode: String = ""
    public var newPassword: String = ""
    /// Code de vérification de l'email (normalisé en chiffres par la vue).
    public var verifyCode: String = ""

    public private(set) var isSubmitting = false
    public private(set) var errorMessage: String?
    public private(set) var infoMessage: String?

    private let repository: AuthRepository

    /// Fournit un ID token Google (SDK Google Sign-In, câblé par l'app).
    /// `nil` → le bouton Google est masqué.
    public var googleIdTokenProvider: (@MainActor () async throws -> String)?

    /// Appelé après une connexion réussie (enregistrement du jeton push, préchargements…).
    public var onSignedIn: (@MainActor () -> Void)?

    /// - Parameter repository: accès REST à `/auth/…`.
    public init(repository: AuthRepository) {
        self.repository = repository
    }

    // MARK: - Validation

    /// Le formulaire courant est-il soumettable ? (règles alignées sur le backend).
    public var canSubmit: Bool {
        guard !isSubmitting else { return false }
        switch mode {
        case .login:
            return email.contains("@") && password.count >= 8
        case .register:
            // Étape 1 : email + mot de passe + confirmation uniquement. Les autres infos
            // (nom, pays, numéro) sont demandées APRÈS validation du code email.
            return email.contains("@") && password.count >= 8 && password == confirmPassword
        case .forgot:
            return email.contains("@")
        case .reset:
            return email.contains("@") && (4...8).contains(resetCode.count) && newPassword.count >= 8
        }
    }

    /// Change de mode (connexion/inscription/oublié/reset), en nettoyant les messages.
    public func setMode(_ newMode: AuthMode) {
        mode = newMode
        errorMessage = nil
        infoMessage = nil
    }

    /// Efface les messages affichés (appelé à chaque édition de champ).
    public func clearMessages() {
        errorMessage = nil
        infoMessage = nil
    }

    // MARK: - Cycle de vie

    /// Réhydrate la session au lancement (via le refresh token persisté au Keychain).
    public func bootstrap() async {
        do {
            let me = try await repository.me()
            authState = .signedIn(user: me.user)
            onSignedIn?()
        } catch {
            authState = .signedOut
        }
    }

    // MARK: - Actions

    /// Soumet le formulaire courant selon ``mode``.
    public func submit() async {
        guard canSubmit else { return }
        isSubmitting = true
        errorMessage = nil
        infoMessage = nil
        defer { isSubmitting = false }

        switch mode {
        case .login:
            do {
                let session = try await repository.login(email: email.trimmed, password: password)
                authState = .signedIn(user: session.user)
                onSignedIn?()
            } catch {
                errorMessage = friendlyMessage(error)
            }

        case .register:
            // Le backend envoie automatiquement le code de vérification par email.
            do {
                let session = try await repository.register(
                    RegisterRequest(
                        email: email.trimmed,
                        password: password,
                        referralCode: referralCode.trimmed.uppercased().nilIfBlank
                    )
                )
                authState = .pendingEmailVerification(user: session.user, email: email.trimmed)
            } catch {
                errorMessage = friendlyMessage(error)
            }

        case .forgot:
            do {
                try await repository.forgotPassword(email: email.trimmed)
                mode = .reset
                infoMessage = AppStrings.current.resetCodeSent
            } catch {
                errorMessage = friendlyMessage(error)
            }

        case .reset:
            do {
                try await repository.resetPassword(
                    email: email.trimmed,
                    code: resetCode.digitsOnly,
                    newPassword: newPassword
                )
                mode = .login
                password = ""
                infoMessage = AppStrings.current.passwordResetDone
            } catch {
                errorMessage = friendlyMessage(error)
            }
        }
    }

    /// Connexion Google native : récupère l'ID token via le SDK puis l'échange côté backend.
    public func signInWithGoogle() async {
        guard let provider = googleIdTokenProvider else { return }
        isSubmitting = true
        errorMessage = nil
        defer { isSubmitting = false }
        do {
            let idToken = try await provider()
            let session = try await repository.loginWithGoogle(idToken: idToken)
            authState = .signedIn(user: session.user)
            onSignedIn?()
        } catch {
            errorMessage = (error as? APIError)?.message ?? AppStrings.current.errGoogleSignInFailed
        }
    }

    /// Vérifie le code email ; en cas de succès, passe à la saisie des infos de profil.
    public func verifyEmail() async {
        guard case let .pendingEmailVerification(user, _) = authState else { return }
        guard (4...8).contains(verifyCode.count) else { return }
        isSubmitting = true
        errorMessage = nil
        defer { isSubmitting = false }
        do {
            try await repository.verifyEmailOtp(code: verifyCode.digitsOnly)
            authState = .pendingProfileCompletion(user: user)
        } catch {
            errorMessage = friendlyMessage(error)
        }
    }

    /// (Re)envoie le code de vérification email.
    public func resendEmailCode() async {
        do {
            try await repository.sendEmailOtp()
            infoMessage = AppStrings.current.newCodeSent
        } catch {
            errorMessage = friendlyMessage(error)
        }
    }

    /// Passe la vérification (non bloquante) → saisie des infos de profil.
    public func skipVerification() {
        guard case let .pendingEmailVerification(user, _) = authState else { return }
        authState = .pendingProfileCompletion(user: user)
    }

    /// Étape 3 : enregistre les informations de profil (nom, pays, numéro) puis entre dans
    /// l'app. Le numéro est optionnel ; le nom est requis.
    public func completeProfile() async {
        guard case let .pendingProfileCompletion(user) = authState else { return }
        guard !fullName.trimmed.isEmpty else {
            errorMessage = AppStrings.current.errNameRequired
            return
        }
        isSubmitting = true
        errorMessage = nil
        defer { isSubmitting = false }
        do {
            try await repository.updateProfile(
                UpdateProfileRequest(
                    fullName: fullName.trimmed,
                    countryCode: country.code,
                    phone: phone.nilIfBlank,
                    phoneCountryCode: country.dial
                )
            )
            authState = .signedIn(user: user)
            onSignedIn?()
        } catch {
            errorMessage = friendlyMessage(error)
        }
    }

    /// Passe la saisie des infos de profil (complétables plus tard depuis le profil).
    public func skipProfile() {
        guard case let .pendingProfileCompletion(user) = authState else { return }
        authState = .signedIn(user: user)
        onSignedIn?()
    }

    /// Déconnexion : invalide la session serveur, efface le Keychain, réinitialise le form.
    public func signOut() async {
        await repository.logout()
        authState = .signedOut
        resetForm()
    }

    // MARK: - Interne

    private func resetForm() {
        mode = .login
        email = ""
        password = ""
        confirmPassword = ""
        fullName = ""
        phone = ""
        country = Countries.default
        referralCode = ""
        resetCode = ""
        newPassword = ""
        verifyCode = ""
        errorMessage = nil
        infoMessage = nil
    }

    /// Traduit une erreur backend en message utilisateur (mêmes cas qu'Android).
    private func friendlyMessage(_ error: Error) -> String {
        let s = AppStrings.current
        guard let api = error as? APIError else { return s.errGeneric }
        switch true {
        case api.httpStatus == 401: return s.errWrongCredentials
        case api.httpStatus == 409: return s.errEmailTaken
        case api.code == ErrorCode.otpInvalid, api.code == ErrorCode.otpExpired: return s.errCodeInvalid
        case api.code == ErrorCode.validation: return s.errInvalidFields
        case api.isRetriable: return s.errUnstableConnection
        default: return api.message
        }
    }
}
